[XLA:CPU] [oneDNN] Adding support for MatMul + BiasAdd fusion with oneDNN custom call rewrite

mahmoud-abuzaina opened the pull request:
This PR enables OneDNN library call for the matched XLA HLO MatMul + BiasAdd patterns through custom_call instruction. In particular, this PR:

1. Renames oneDNN matmul rewriter pass and moves it after layout assignment pass.
2. Enables transposed input support for oneDNN matmul
3. Enables MatMul + BiasAdd pattern matching with oneDNN rewrite
4. Adds multiple tests with different variations of MatMul + BiasAdd pattern

Original Tensorflow PR: https://github.com/tensorflow/tensorflow/pull/62243
With contribution from @akhilgoe

penpornk reviewed (approved):
Approving since @d0k already approved the original PR.

penpornk reviewed (changes requested):
Thank you for adding many test cases!

mahmoud-abuzaina commented on xla/service/cpu/backend_config.proto:
```diff
@@ -23,5 +26,5 @@ message OneDnnMatMulConfig {
     GELU_ERF = 4;
     GELU_TANH = 5;
   }
-  repeated FusionKind fused_ops = 3;
```
Fixed.

mahmoud-abuzaina commented on xla/service/cpu/cpu_compiler.cc:
```diff
@@ -892,6 +884,14 @@ Status CpuCompiler::RunHloPassesAfterLayoutAssn(
+    pipeline.AddPass<OneDnnMatMulRewriter>();
+    pipeline.AddPass<ReshapeDecomposer>();
+  }
+#endif  // INTEL_MKL && ENABLE_ONEDNN_V3
```
Right, I think we just need once before the Matmul pass to simplify the shape. I removed the redundant one.

mahmoud-abuzaina commented on xla/service/cpu/onednn_matmul_rewriter.cc:
```diff
@@ -0,0 +1,396 @@
+}
+
+template <typename Pattern>
+auto IntermediateAllowedInstructions(HloInstruction** bitcast,
```
Done.

mahmoud-abuzaina commented on xla/service/cpu/onednn_matmul_rewriter.cc:
```diff
@@ -0,0 +1,396 @@
+  });
+}
+
+auto ConvertPattern(HloInstruction** instr) {
```
Done.

mahmoud-abuzaina commented on xla/service/cpu/onednn_matmul_rewriter.cc:
```diff
@@ -0,0 +1,396 @@
+}
+
+// We also check if the convert instruction has only one use.
+bool AllOperandsConvertedFromBF16ToF32(const HloInstruction* instr) {
```
I tagged few those functions as inline.

mahmoud-abuzaina commented on xla/service/cpu/onednn_matmul_rewriter.cc:
```diff
@@ -0,0 +1,396 @@
+        new_operands.push_back(operand);
+      }
+
+      // For addened, match one of these patterns:
```
Fixed.

mahmoud-abuzaina commented on xla/service/cpu/onednn_matmul_rewriter.cc:
```diff
@@ -0,0 +1,396 @@
+      // For addened, match one of these patterns:
+      // 1. any(addend) -> broadcast -> add
+      // 2. any(addend) -> bitcast -> add
+      bool bias_broad = Match(addend, nonscalar_broadcast);
```
Done.

mahmoud-abuzaina commented on xla/service/cpu/onednn_matmul_rewriter.cc:
I tried to use "git mv" earlier, it did not work because the changes are relatively big in that file so git insists to mark the relevant files as deleted/new. I also tried (after your review) to do rename & code changes in two separate commits to force git to recognize the rename but it still marked them as new/deleted eventually. I will give it another try and push another commit if it works.

mahmoud-abuzaina replied:
I could not fix this branch probably because of the commits in between. I created a replacement PR with same changes https://github.com/openxla/xla/pull/7681, even though git still shows deleted/new file in the new PR diff but history seems to be still maintained for this file.

penpornk replied:
Got it. Thank you very much for your efforts, @mahmoud-abuzaina! I've approved PR  https://github.com/openxla/xla/pull/7681.

mahmoud-abuzaina commented on xla/service/cpu/onednn_matmul_rewriter.cc:
```diff
@@ -0,0 +1,396 @@
+  should_rewrite &= LayoutUtil::IsMonotonicWithDim0Major(rhs_shape.layout());
+  if (!should_rewrite) return false;
+  should_rewrite &= LayoutUtil::IsMonotonicWithDim0Major(output_shape.layout());
+  if (!should_rewrite) return false;
```
Done.

mahmoud-abuzaina commented on xla/service/cpu/onednn_matmul_rewriter.cc:
```diff
@@ -0,0 +1,396 @@
+    should_rewrite &= num_mac_ops >= (1 << 23);
+  } else {
+    should_rewrite &= num_mac_ops >= (1 << 18);
+  }
```
Done.

mahmoud-abuzaina commented on xla/service/cpu/onednn_matmul_rewriter.cc:
```diff
@@ -0,0 +1,396 @@
+    OneDnnMatMulConfig* matmul_config =
+        backend_config.mutable_onednn_matmul_config();
+    bool transpose_a = (lhs_dim_k == lhs_shape.rank() - 1) ? false : true;
+    bool transpose_b = (rhs_dim_k == rhs_shape.rank() - 2) ? false : true;
```
Done.

mahmoud-abuzaina commented on xla/service/cpu/onednn_matmul_rewriter.cc:
```diff
@@ -0,0 +1,396 @@
+      HloInstruction* new_instr;
+      // If matched pattern has custom-call -> bitcast -> add, then we need to
+      // insert bitcast after the new fusion to maintain the correct shape
+      // (new-custom-call -> bitcast). Also, this will be followd by -> convert
```
Fixed.

mahmoud-abuzaina commented on xla/service/cpu/onednn_matmul_rewriter.cc:
```diff
@@ -0,0 +1,396 @@
+class OneDnnMatMulRewriteVisitor : public DfsHloRewriteVisitor {
+ public:
+  // Matches patterns for possible MatMul fusions that are supported by oneDNN
+  // library. Matched hlo instruction(s) are replaced by custom call.
```
Fixed.

mahmoud-abuzaina commented on xla/service/cpu/onednn_matmul_rewriter.h:
```diff
@@ -0,0 +1,48 @@
+namespace xla {
+namespace cpu {
+
+// This pass pattern-matches hlo dot instructions and rewrites into custom
```
Fixed.

mahmoud-abuzaina commented on xla/service/cpu/onednn_util.h:
```diff
@@ -0,0 +1,45 @@
+namespace cpu {
+
+inline bool IsSupportedType(xla::PrimitiveType dtype) {
+  using namespace tsl::port;
```
Fixed by resolving merge conflicts.

mahmoud-abuzaina commented on xla/service/cpu/onednn_util.h:
Done.

mahmoud-abuzaina commented on xla/tests/onednn_matmul_test.cc:
```diff
@@ -17,18 +17,29 @@ limitations under the License.
-class MatmulTest : public HloTestBase {};
+class MatmulTest : public HloTestBase {
+ protected:
+  const char* fused_matmul_bias = R"(
```
Fixed.

mahmoud-abuzaina commented on xla/tests/onednn_matmul_test.cc:
```diff
@@ -77,6 +98,161 @@ TEST_F(MatmulTest, SimpleTestF32TransposeB) {
+  ; CHECK-DAG:       "fused_ops":["BIAS"]
+  ; CHECK-DAG:   }
+  ; CHECK:     }
+  )");
```
Done.

mahmoud-abuzaina commented:
Thank you @penpornk for the review. I have addressed your comments except for the "git mv" one. Please see my comment above about that. If I fix it I will push another commit.
